packages/multichain/src/caip25Permission.ts

```diff
@@ -0,0 +1,283 @@
+
+import { getEthAccounts } from './adapters/caip-permission-adapter-eth-accounts';
+import { assertScopesSupported } from './scope/assert';
+import { validateAndNormalizeScopes } from './scope/authorization';
+import type {
+  ExternalScopeString,
+  InternalScopeObject,
+  InternalScopesObject,
+} from './scope/types';
+
+export type Caip25CaveatValue = {
+  requiredScopes: InternalScopesObject;
+  optionalScopes: InternalScopesObject;
+  sessionProperties?: Record<string, Json>;
+  isMultichainOrigin: boolean;
+};
+
+export const Caip25CaveatType = 'authorizedScopes';
+
+export const createCaip25Caveat = (value: Caip25CaveatValue) => {
+  return {
+    type: Caip25CaveatType,
+    value,
+  };
+};
+
+export const Caip25EndowmentPermissionName = 'endowment:caip25';
+
+type Caip25EndowmentSpecification = ValidPermissionSpecification<{
+  permissionType: PermissionType.Endowment;
+  targetName: typeof Caip25EndowmentPermissionName;
+  endowmentGetter: (_options?: EndowmentGetterParams) => null;
+  validator: PermissionValidatorConstraint;
+  allowedCaveats: Readonly<NonEmptyArray<string>> | null;
+}>;
+
+type Caip25EndowmentSpecificationBuilderOptions = {
+  methodHooks: {
+    findNetworkClientIdByChainId: (chainId: Hex) => NetworkClientId;
+    listAccounts: () => { address: Hex }[];
+  };
+};
+
+/**
+ * Helper that returns a `endowment:caip25` specification that
+ * can be passed into the PermissionController constructor.
+ *
+ * @param builderOptions - The specification builder options.
+ * @param builderOptions.methodHooks - The RPC method hooks needed by the method implementation.
+ * @returns The specification for the `caip25` endowment.
+ */
+const specificationBuilder: PermissionSpecificationBuilder<
+  PermissionType.Endowment,
+  Caip25EndowmentSpecificationBuilderOptions,
+  Caip25EndowmentSpecification
+> = ({ methodHooks }: Caip25EndowmentSpecificationBuilderOptions) => {
+  return {
+    permissionType: PermissionType.Endowment,
+    targetName: Caip25EndowmentPermissionName,
+    allowedCaveats: [Caip25CaveatType],
+    endowmentGetter: (_getterOptions?: EndowmentGetterParams) => null,
+    validator: (permission: PermissionConstraint) => {
+      const caip25Caveat = permission.caveats?.[0];
+      if (
+        permission.caveats?.length !== 1 ||
+        caip25Caveat?.type !== Caip25CaveatType
+      ) {
+        throw new Error(
+          `${Caip25EndowmentPermissionName} error: Invalid caveats. There must be a single caveat of type "${Caip25CaveatType}".`,
+        );
+      }
+
+      const { requiredScopes, optionalScopes, isMultichainOrigin } =
+        caip25Caveat.value as Caip25CaveatValue;
+
+      if (
+        !requiredScopes ||
+        !optionalScopes ||
+        typeof isMultichainOrigin !== 'boolean'
+      ) {
+        throw new Error(
+          `${Caip25EndowmentPermissionName} error: Received invalid value for caveat of type "${Caip25CaveatType}".`,
+        );
+      }
```

Comment:
We're skipping a lot of validation steps here by casting to `Caip25CaveatValue`. We'd be letting through an invalid caveat shape here.

Here's a suggested step forward:

```suggestion
      if (
        !caip25Caveat.value ||
        !hasProperty(caip25Caveat.value, 'requiredScopes') ||
        !hasProperty(caip25Caveat.value, 'optionalScopes') ||
        !hasProperty(caip25Caveat.value, 'isMultichainOrigin') ||
        typeof caip25Caveat.value.isMultichainOrigin !== 'boolean'
      ) {
        throw new Error(
          `${Caip25EndowmentPermissionName} error: Received invalid value for caveat of type "${Caip25CaveatType}".`,
        );
      }

      const { requiredScopes, optionalScopes } = caip25Caveat.value;
```

However this still leaves `requiredScopes` and `optionalScopes` as type `unknown`, we'd need to validate that they have the correct properties as well. `validateAndNormalizeScopes` only validates beyond what the type guarantees, we don't have a method for validating that it matches the type.

Reply:
Done here: https://github.com/MetaMask/core/pull/4784/commits/386ed028379603bda9b65b0e425bb5722f010a6a

Reply:
done here: https://github.com/MetaMask/core/pull/4784/commits/d10f7d04fca83cd8d79a58919c838597285c190b